packages/utils/src/stableMath.ts

```diff
@@ -0,0 +1,209 @@
+import { BigNumber } from "bignumber.js";
+
+const BIG_ZERO = new BigNumber(0)
+const BIG_ONE = new BigNumber(1)
+const BIG_TEN = new BigNumber(10)
+
+
+
+// Computes the invariant given the current balances, using the Newton-Raphson approximation.
+// The amplification parameter equals: A n^(n-1)
+export const _calculateInvariant = (
+  amplificationParameter: BigNumber,
+  balances: BigNumber[]
+) :BigNumber => {
+    /**********************************************************************************************
+    // invariant                                                                                 //
+    // D = invariant                                                  D^(n+1)                    //
+    // A = amplification coefficient      A  n^n S + D = A D n^n + -----------                   //
+    // S = sum of balances                                             n^n P                     //
+    // P = product of balances                                                                   //
+    // n = number of tokens                                                                      //
+    *********x************************************************************************************/
+    // D(n+1) = D(n) - (f(Dn)/f'(Dn)) = (AnnS + nDp)Dn / [(Ann-1)Dn + (n+1)Dp]
+    // We support rounding up or down.
+
+    let sum = BIG_ZERO;
+    let numTokens = balances.length;
+    for (let i = 0; i < numTokens; i++) {
+        sum = sum.plus(balances[i])
+    }
+   
+    if (sum.isEqualTo(0)) {
+        return BIG_ZERO;
+    }
+
+    let Dprev = BIG_ZERO;  // Dn
+    let D = sum;    // Dn+1
+    let Ann = new BigNumber(amplificationParameter).times(numTokens);  // An
+    
+    for (let i = 0; i < 255; i++) {
+        let D_P = D;
+
+        // calculate derivative of f(D)
+        for(let j = 0; j < numTokens; j++) {
+          D_P = D_P.multipliedBy(D).dividedBy(new BigNumber(balances[j]).times(numTokens));
+        }
+       
+        Dprev = D;
+        
+        //D = (Ann * sum + D_P * numTokens) * D / ((Ann - 1) * D + (numTokens + 1) * D_P)
+        let temp1 = (Ann.multipliedBy(sum)).plus(D_P.times(numTokens))
+        let temp2 = ((Ann.minus(BIG_ONE)).multipliedBy(Dprev)).plus(D_P.times(numTokens + 1))
+        D = temp1.multipliedBy(Dprev).dividedBy(temp2)
+        
+        if (D.gt(Dprev)) {
+          if (D.minus(Dprev).lte(BIG_ONE)) {
+            break;
+          }
+        } else {
+          if(Dprev.minus(D).lte(BIG_ONE)) {
+            break;
+          }
+        }
```

Comment:
Seems like you can take the absolute value of `D.minus(Dprev).abs()` and it'll be a bit easier to read